packages/core/src/fhir-deduplication/deduplicate-fhir.ts

```diff
@@ -29,350 +30,285 @@ import { createRef } from "./shared";
+    ...medRequestResult.refReplacementMap,
+    ...medStatementResult.refReplacementMap,
+    ...practitionersResult.refReplacementMap,
+    ...conditionsResult.refReplacementMap,
+    ...allergiesResult.refReplacementMap,
+    ...encountersResult.refReplacementMap,
+    ...diagReportsResult.refReplacementMap,
+    ...immunizationsResult.refReplacementMap,
+    ...proceduresResult.refReplacementMap,
+    ...obsSocialResult.refReplacementMap,
+    ...obsLabsResult.refReplacementMap,
+    ...obsVitalsResult.refReplacementMap,
+    ...obsOthersResult.refReplacementMap,
+    ...locationsResult.refReplacementMap,
+    ...organizationsResult.refReplacementMap,
+    ...relatedPersonsResult.refReplacementMap,
+    ...famMemHistoriesResult.refReplacementMap,
+    ...coveragesResult.refReplacementMap,
+  ]);
+
+  resourceArrays = replaceResourceReferences(resourceArrays, combinedReplacementMap);
+
+  // Remove resources with dangling links
+  const { updatedResourceArrays, deletedRefs } = removeResourcesWithDanglingLinks(
+    resourceArrays,
     danglingLinks
   );
+  resourceArrays = updatedResourceArrays;
 
-  const compositionsNoDangling = removeResourcesWithDanglingLinks(
-    resourceArrays.compositions,
-    danglingLinks
+  const { updatedResourceArrays: updatedResourceArrays2 } = removeResourcesWithDanglingLinks(
+    resourceArrays,
+    deletedRefs
   );
+  resourceArrays = updatedResourceArrays2;
 
-  const deduplicatedBundle: Bundle = cloneDeep(fhirBundle);
-  deduplicatedBundle.entry = [...deduplicatedNoDangling, ...compositionsNoDangling].map(
-    buildBundleEntry
-  );
+  const deduplicatedEntries: BundleEntry<Resource>[] = [];
+  // Rebuild the entries with deduplicated resources and add whatever is left unprocessed
+  for (const [, resources] of Object.entries(resourceArrays)) {
+    const entriesArray = resources && Array.isArray(resources) ? resources : [resources];
+    const entriesFlat = entriesArray.flatMap(v => v || []);
+    const entriesWithDeduplicatedReferences = entriesFlat.map(removeDuplicateReferences);
+    deduplicatedEntries.push(...entriesWithDeduplicatedReferences);
+  }
 
-  deduplicatedBundle.total = deduplicatedNoDangling.length;
+  // Rebuild the final bundle
+  const deduplicatedBundle: Bundle = cloneDeep(fhirBundle);
+  deduplicatedBundle.entry = deduplicatedEntries.map(entry => buildBundleEntry(entry as Resource));
+  deduplicatedBundle.total = deduplicatedBundle.entry.length;
 
   return deduplicatedBundle;
 }
 
-/**
- * Finds and updates references to the deduplicated resources
- */
-function replaceResourceReferences(
-  resourceArrays: ExtractedFhirTypes,
-  idMap: Map<string, string[]>
-): ExtractedFhirTypes {
-  let updatedArrays = JSON.stringify(resourceArrays);
-  for (const [masterRef, consumedRefs] of idMap.entries()) {
-    for (const consumedRef of consumedRefs) {
-      const regex = new RegExp(consumedRef, "g");
-      updatedArrays = updatedArrays.replace(regex, masterRef);
```

Comment:
runtime of this is v bad. regexReplace is O(s) where s is the size of the entire set of patient data as a string, and then performing that n times for every k entry in idMap, where n is the number of links in the string[] and k is the number of entries. 

O(s*n*k) 

and s here is way larger than the number of entries in s, since s is a string

Reply:
replacement function uses a hashMap and iterates through the total number of entries. If there are e entries, runtime is 
O(e)